flocker/control/_diffing.py

```diff
@@ -0,0 +1,247 @@
+            applied.
+        """
+        pass
+
+
+@implementer(_DiffChange)
+class _Remove(PClass):
+    """
+    A ``_DiffChange`` that removes an object from a ``PSet`` or a key from a
+    ``PMap`` inside a nested object tree.
+
+    :ivar path: The path in the nested object tree of the object to be removed
+        from the import set.
+    """
+
+    path = field(
+        type=PVector,
+        factory=freeze
+    )
+
+    def apply(self, obj):
+        obj_path = self.path[:-1]
+        removal_path = self.path[-1]
+        return obj.transform(obj_path, lambda o: o.remove(removal_path))
+
+
+@implementer(_DiffChange)
+class _Set(PClass):
+    """
+    A ``_DiffChange`` that sets a field in a ``PClass`` or sets a key in a
+    ``PMap``.
+
+    :ivar path: The path in the nested object to the field/key to be set to a
+        new value.
+
+    :ivar val: The value to set the field/key to.
+    """
+    path = field(
+        type=PVector,
+        factory=freeze
+    )
+    val = field()
+
+    def apply(self, obj):
+        return obj.transform(self.path, self.val)
+
+
+@implementer(_DiffChange)
+class _Add(PClass):
+    """
+    A ``_DiffChange`` that adds an item to a ``PSet``.
+
+    :ivar path: The path to the set to which the item will be added.
+
+    :ivar item: The item to be added to the set.
+    """
+    path = field(
+        type=PVector,
+        factory=freeze
+    )
+    item = field()
+
+    def apply(self, obj):
+        return obj.transform(self.path, lambda x: x.add(self.item))
+
+
+@implementer(_DiffChange)
+class _Diff(PClass):
+    """
+    A ``_DiffChange`` that is simply the serial application of other diff
+    changes.
+
+    This is the object that external modules get and use to apply diffs to
+    objects.
+
+    :ivar changes: A vector of ``_DiffChange`` s that represent a diff between
+        two objects.
+    """
+
+    changes = field(
+        type=PVector,
+        factory=freeze
+    )
+
+    def apply(self, obj):
+        for c in self.changes:
+            obj = c.apply(obj)
+        return obj
+
+
+def _create_diffs_for_sets(current_path, set_a, set_b):
+    """
+    Computes a series of ``_DiffChange`` s to turn ``set_a`` into ``set_b``
+    assuming that these sets are at ``current_path`` inside a nested pyrsistent
+    object.
+
+    :param current_path: And iterable of pyrsistent object describing the path
+        inside the root pyrsistent object where the other arguments are
+        located.  See ``PMap.transform`` for the format of this sort of path.
+
+    :param set_a: The desired input set.
+
+    :param set_b: The desired output set.
+
+    :returns: An iterable of ``_DiffChange`` s that will turn ``set_a`` into
+        ``set_b``.
+    """
+    resulting_diffs = pvector([]).evolver()
+    for item in set_a.difference(set_b):
+        resulting_diffs.append(
+            _Remove(path=current_path.append(item))
+        )
```

Comment:
:question: Might be neat if `_Remove` and `_Add` had the same signature. 
`_Remove(path=current_path, item=item)` should work I think.